### PR TITLE
[ROCm][CI] periodic-rocm-mi350: run >2-GPU distributed tests via multigpu marker

### PR DESCRIPTION
**Pull request by amdfaa**
## Motivation

The `periodic-rocm-mi350` workflow runs on `gfx950.4` (4-GPU) runners, but the standard `distributed` job runs on 2-GPU runners and therefore silently skips every distributed test that needs more than 2 GPUs (e.g. 2D FSDP+TP, ND device-mesh, larger-scale NCCL). This PR routes exactly those >2-GPU tests onto the 4-GPU runners, without re-running the 1-2 GPU / CPU-gloo low-world-size tests that already run on the standard job, so the extra GPUs add real coverage rather than duplicate work.

## Summary

Enable the `periodic-rocm-mi350` workflow to run the distributed tests that need more GPUs than the standard 2-GPU distributed runners provide on the `gfx950.4` (4-GPU) runners, while NOT re-running the 1-2 GPU / CPU-gloo low-world-size tests that already run on the standard `distributed` job.

This uses main's **native `multigpu` marker / `--multigpu-filter` machinery** rather than a custom collection plugin:

- `pytest_itemcollected` in `test/conftest.py` already tags every process-spawning distributed test with the `multigpu` marker. This adds a `multigpu_extra` tier for tests whose resolved accelerator requirement exceeds the standard 2-GPU runner. The requirement is resolved at collection time by **combining two signals and taking the larger**: the `skip_if_lt_x_gpu` / `requires_world_size` decorator stamp and the multi-process base-class `world_size`. A test is selected when `max(decorator, world_size) > STANDARD_DISTRIBUTED_GPUS (2)`, so e.g. a test carrying only `skip_if_lt_x_gpu(2)` but whose class hardcodes `world_size = 4` is correctly routed to the 4-GPU tier.
- The selection has **no upper GPU cap**: it selects *any* test needing more than 2 GPUs. Tests needing more than 4 GPUs are still selected but runtime-skip on a 4-GPU box (via their own `skip_if_lt_x_gpu`), so no coverage is silently dropped.
- Two gates keep non-GPU-scaling tests out of the tier:
  - **CPU/gloo gate:** a test whose file targets a CPU transport (e.g. `test_c10d_gloo`) spawns ranks, not GPUs, so it never qualifies. The file identity is resolved from the collected node (`path` / `fspath` / `nodeid`), which is reliable under `run_test.py`, and backend names are matched as whole `_`-delimited tokens.
  - **LocalTensor simulation exclusion:** `*WithLocalTensor` classes inherit a multi-process base and a `world_size` of 4 but run single-process under `LocalTensorMode` on one GPU, so their simulated mesh size is not a GPU count; they are detected and excluded from both markers.
  - On any ambiguity (unresolvable `world_size`) the resolver favors coverage and routes the test to the larger runner.
- `skip_if_lt_x_gpu` stamps `_min_gpus_required` (mirroring `requires_world_size._required_world_size`) so the requirement resolves without running the test. `STANDARD_DISTRIBUTED_GPUS = 2` lives in `common_distributed.py`.
- `run_test.py` gains the `multigpu-extra` choice for `--multigpu-filter` (maps to `-m "multigpu_extra"`).
- `.ci/pytorch/test.sh` gains a `distributed_4gpu` config that runs the sharded python distributed suite with that filter (python-only; the multi-GPU C++/mpiexec tests already run on the standard `distributed` job).
- The ROCm `HIP_VISIBLE_DEVICES=0,1,2,3` override for the `distributed` config is removed; `DEFAULT_WORLD_SIZE` already caps ROCm world size at 4 via `device_count`, so the override was redundant (and forced 4 visible devices even on 2-GPU nodes).
- `periodic-rocm-mi350.yml` is re-enabled (schedule + `ciflow/periodic-rocm-mi350` push tag) with a 5-shard `distributed_4gpu` matrix on `gfx950.4`; `arc.yaml` and `pytorch-probot.yml` register the runner label and ciflow tag.

The mechanism is device-generic: CUDA multi-GPU jobs can reuse `--multigpu-filter multigpu-extra` too. This is an **alternative implementation** to #187450 (which used a custom `MinGpuFilterPlugin` gated by `PYTORCH_TEST_MIN_GPU`), offered for comparison of the two approaches.

## Test plan

- Device-generic resolver unit test (runs in CI where torch is built):
```
python test/run_test.py -i distributed/test_multigpu_marker --verbose
```
- Syntax / config checks (torch not built in the dev env):
```
python -m py_compile test/conftest.py test/run_test.py torch/testing/_internal/common_distributed.py test/distributed/test_multigpu_marker.py
bash -n .ci/pytorch/test.sh
```

**CI: `periodic-rocm-mi350` [run 30127795566](https://github.com/pytorch/pytorch/actions/runs/30127795566) (head `b27a5c446b9f`) is green across all 5 `distributed_4gpu` shards on `linux.rocm.gpu.gfx950.4`:**

| shard | conclusion | duration |
|-------|------------|----------|
| [distributed_4gpu 1/5](https://github.com/pytorch/pytorch/actions/runs/30127795566/job/89599321396) | success | 1h10m |
| [distributed_4gpu 2/5](https://github.com/pytorch/pytorch/actions/runs/30127795566/job/89599321437) | success | 1h26m |
| [distributed_4gpu 3/5](https://github.com/pytorch/pytorch/actions/runs/30127795566/job/89599321434) | success | 1h28m |
| [distributed_4gpu 4/5](https://github.com/pytorch/pytorch/actions/runs/30127795566/job/89599321464) | success | 1h02m |
| [distributed_4gpu 5/5](https://github.com/pytorch/pytorch/actions/runs/30127795566/job/89599321426) | success | 1h34m |

Aggregate over the run: 2149 distinct methods selected (1689 executed, 460 runtime-skipped), **0 failures**. At the JUnit `<testcase>` level (parametrized variants included): 3022 entries, 2156 passed, 866 skipped, 0 failed.

Representative evidence that the >2-GPU tests actually ran on 4 GPUs (PASSED):
```
distributed/_composable/fsdp/test_fully_shard_state_dict.py::TestFullyShardStateDictMultiProcess::test_2d_state_dict_correctness
distributed/_composable/fsdp/test_fully_shard_state_dict.py::TestFullyShardStateDictMultiProcess::test_dp_tp_state_dict_save_load
distributed/checkpoint/test_fsdp_tp_checkpoint_conversion.py::TestFsdpTpCheckpointConversion::test_fsdp_to_tp
```
Representative >4-GPU tests correctly selected then runtime-skipped on the 4-GPU box (no coverage silently dropped):
```
SKIPPED [Need at least 8] distributed/_composable/fsdp/test_fully_shard_training.py::TestFullyShardNDTraining::test_shard_placement_fn_tp_ep
SKIPPED [Need at least 8] distributed/test_c10d_nccl.py::ProcessGroupNCCLLargerScaleTest::test_comm_split_group_larger_scale
SKIPPED [Need at least 6] distributed/tensor/test_redistribute.py::UnevenFlattenedReduceScatterTest::test_partial_partial_to_shard_shard_uneven
```

- Empirical selection validation (post-fix vs earlier pre-fix run):
  - The 5 genuine 4-GPU tests that were previously missed now run and pass on the 4-GPU box: `TestFullyShard2DTraining::{test_tp_with_fsdp_offloading, test_train_parity_2d_mlp, test_train_parity_2d_transformer}` and `TestDistributedFailure::{test_dummy_reader_works, test_dummy_writer_works}`.
  - CPU/gloo and LocalTensor-simulation over-selection is eliminated (150 gloo -> 0, 422 LocalTensor sims -> 0). Among executed tests, <=2-GPU over-selection drops from 32.9% to 2.7% (residual is device_count-scaling / unresolvable-`world_size` favor-coverage cases).

Note: the resolver unit test `test/distributed/test_multigpu_marker.py` does not appear in the `distributed_4gpu` artifacts because it is a plain `TestCase` (not `multigpu`-marked) and is excluded by `-m multigpu`; it runs under the standard `distributed` config.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

Made with Cursor

### CODE REVIEW

File: test/conftest.py [hunk 1]
```diff
@@ -381,15 +381,137 @@ def _spawns_multiple_processes(
...
+    the 3-4 GPU tests with `--multigpu-filter multigpu-extra`.
     """
-    if _spawns_multiple_processes(getattr(item, "cls", None)):
+    cls = getattr(item, "cls", None)
+    if _is_local_tensor_simulation(cls):
+        return
+    if _spawns_multiple_processes(cls):
         item.add_marker("multigpu")
+        if _needs_extra_gpus(item, cls):
+            item.add_marker("multigpu_extra")
```
Comment (jithunnair-amd) [inline test/conftest.py]:
```suggestion
    if _needs_gpus_more_than(item, cls, args.min_gpu_count):
        item.add_marker("multigpu")
```

where

```
def _needs_gpus_more_than(item: Any, cls: type | None, int: min_gpus):
  if _is_local_tensor_simulation(cls):
        return False
  if _spawns_multiple_processes(cls):
    if _is_cpu_backed(item):
        return False
    world_size = _probe_world_size(cls)
    if world_size == 0:
        return True
    dec = _decorator_gpu_requirement(_safe_obj(item))
    return max(dec, world_size) > min_gpus
```

**Comment by jithunnair-amd** [comment]
@claude /pr-review

**Comment by jeffdaily** [comment]
Closing as superseded by #195597, which carries this change forward with the review feedback addressed; this branch could not be updated. Please continue there.

Analysis assisted by AI.
